app/assets/javascripts/toc.js

```diff
@@ -1,28 +1,88 @@
-$(document).ready(function() {
-  let element = $('#toc')
-  if(element.length > 0) {
-    let selectedId = element.data('selected')
-    let baseUrl = element.data('url')
-    let initialUrl = `${baseUrl}?node=${selectedId}&full=true`
-
-    $.getJSON(
-      initialUrl,
-      function(data) {
-        element.tree({
-          data: data
-        });
-      }).done(function() {
-        let selectedNode = element.tree('getNodeById', selectedId);
-        element.tree('selectNode', selectedNode);
-    });
+class TocBuilder {
+  constructor(element) {
+    this.element = $(element)
+    // We need a separate data element that is updated by turbolinks so we always
+    // have the correct selected node id. The toc element itself is permanent (not
+    // updated by turbolinks), and is instead updated by event-triggered javascript.
+    this.dataElement = $(`${element}-data`)
+  }
+
+  get baseUrl() {
+    return '/toc/'
+  }
+
+  get initialUrl() {
+    return `${this.baseUrl}?node=${this.selectedId}&full=true`
+  }
+
+  get selectedId() {
+    return this.dataElement.data('selected')
+  }
+
+  build() {
+    this.setupTree()
+    this.setupEventHandlers()
+  }
 
-    $(element).on(
-      'tree.click',
-      function(event) {
-          let node = event.node
-          let url = `/catalog/${node.id}`
-          window.location.href = url
+  setupEventHandlers() {
+    // Listen for click on the leaf node and follow link to component
+    this.element.on('activate_node.jstree', function (e, data) {
+      let location = `/catalog/${data.node.id}`
+      Turbolinks.visit(location)
+      // Scroll to top of page instead of jumping for better user experience
+      window.scrollTo({ top: 0, behavior: 'smooth' });
+    })
+  }
+
+  setupTree() {
+    const that = this
+    this.element.jstree({
+      'core': {
+        'themes': {
+          'icons': false,
+          'dots': false
+        },
+        'data' : {
+          'url': function (node) {
+            if (node.id === '#') {
+              return that.initialUrl
+            } else {
+              return `${that.baseUrl}?node=${node.id}`
+            }
+          }
+        }
       }
-    );
+    });
+  }
+}
+
+
+// Setup table of contents on initial page load
+$(document).ready(function() {
+  const toc = new TocBuilder('#toc')
+  toc.build()
+  // Set initial toc build flag
+  window.buildToc = true
```

Comment:
Suggestion: Instead of just a flag maybe put the `toc` variable in the global space and check to see if it's defined?

Reply:
Are you suggesting that we define buildToc outside of this function/class?

```
let buildToc = true
class TocBuilder {...
```

Reply:
No, I'm suggesting `window.toc = new TocBuilder('#toc')`

Reply:
We still need a flag here. Determining if a TocBuilder is defined is not enough. We want to rebuild the TOC when navigating turbolinks via history (forward /back) but NOT after clicking a link and triggering a turbolinks load.